Create a Wavefront .OBJ file exporter for Triangle Geometries

imikejackson opened the pull request.

joeykleingers committed:
Adding experimental SurfaceMeshToWavefront filter.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Adding missing documentation file.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Fixing copy-and-paste bug.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

imikejackson committed:
Add a license and minor code cleanup

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

JDuffeyBQ commented on DREAM3DReviewFilters/SurfaceMeshToWaveFront.h:
```diff
@@ -0,0 +1,182 @@
+  /**
+   * @brief dataCheck Checks for the appropriate parameter values and availability of arrays
+   */
+  void dataCheck();
```
```suggestion
  void dataCheck() override;
```
Changed to match changes in AbstractFilter.

JDuffeyBQ commented on DREAM3DReviewFilters/SurfaceMeshToWaveFront.cpp:
```diff
@@ -0,0 +1,305 @@
+    ss << "f " << c0 + 1 << " " << c1 + 1 << " " << c2 + 1 << "\n";
+  }
+
+  ss.close();
```
```suggestion

```
std::ofstream is automatically closed in the destructor.

JDuffeyBQ commented on DREAM3DReviewFilters/SurfaceMeshToWaveFront.h:
```diff
@@ -0,0 +1,182 @@
+  using ConstWeakPointer = std::weak_ptr<const Self>;
+  static Pointer NullPointer();
+
+  static std::shared_ptr<SurfaceMeshToWaveFront> New();
```
```suggestion
  static Pointer New();
```
Changed to use typedef.

JDuffeyBQ commented on DREAM3DReviewFilters/SurfaceMeshToWaveFront.cpp:
```diff
@@ -0,0 +1,305 @@
+  FileSystemPathHelper::CheckOutputFile(this, "Output Wavefront File", getOutputWaveFrontFile(), true);
+
+  std::vector<size_t> dims = {1};
+  m_SurfaceMeshNodeTypePtr = getDataContainerArray()->getPrereqArrayFromPath<Int8ArrayType>(this, getSurfaceMeshNodeTypeArrayPath(), dims);
```
SurfaceMeshNodeType isn't used. Should the filter use a DataContainerSelectionFilterParameter instead?

JDuffeyBQ commented on DREAM3DReviewFilters/SurfaceMeshToWaveFront.h:
```diff
@@ -0,0 +1,182 @@
+  PYB11_SHARED_POINTERS(SurfaceMeshToWaveFront)
+  PYB11_FILTER_NEW_MACRO(SurfaceMeshToWaveFront)
+  PYB11_FILTER_PARAMETER(QString, OutputWaveFrontFile)
+  PYB11_FILTER_PARAMETER(DataArrayPath, SurfaceMeshNodeTypeArrayPath)
```
Needs appropriate PYB11_PROPERTY macros.

JDuffeyBQ commented on DREAM3DReviewFilters/SurfaceMeshToWaveFront.cpp:
```diff
@@ -0,0 +1,305 @@
+#include "SIMPLib/FilterParameters/DataContainerSelectionFilterParameter.h"
+#include "SIMPLib/FilterParameters/DataArraySelectionFilterParameter.h"
+#include "SIMPLib/FilterParameters/MultiDataArraySelectionFilterParameter.h"
+#include "SIMPLib/FilterParameters/SeparatorFilterParameter.h"
```
```suggestion

```
Remove unused includes.

JDuffeyBQ commented on Documentation/DREAM3DReviewFilters/SurfaceMeshToWavefront.md:
```diff
@@ -0,0 +1,49 @@
+
+## Description ##
+
+This **Filter** does the following...
```
Documentation needs to be updated.

JDuffeyBQ commented on DREAM3DReviewFilters/SurfaceMeshToWaveFront.cpp:
```diff
@@ -0,0 +1,305 @@
+  size_t numberOfVertices = triangleGeom->getNumberOfVertices();
+
+  std::ofstream ss;
+  ss.open(getOutputWaveFrontFile().toStdString().c_str());
```
```suggestion
  std::ofstream file(getOutputWaveFrontFile().toStdString().c_str());
```
Can combine the open call into the constructor. Rename std::ofstream variable to prevent shadowing of QString below.

imikejackson committed:
Add checks for existing output dir, change variable names

Update the documentation
Add example pipeline
Update filter parameter to use DataContainer selection widget
Add proper python macros.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

JDuffeyBQ commented on DREAM3DReviewFilters/SurfaceMeshToWaveFront.h:
```diff
@@ -49,8 +49,8 @@ class DREAM3DReview_EXPORT SurfaceMeshToWaveFront : public AbstractFilter
-  PYB11_FILTER_PARAMETER(QString, OutputWaveFrontFile)
-  PYB11_FILTER_PARAMETER(DataArrayPath, SurfaceMeshNodeTypeArrayPath)
+  PYB11_PROPERTY(QString, OutputWaveFrontFile)
+  PYB11_PROPERTY(DataArrayPath, TriangleGeometry)
```
```suggestion
  PYB11_PROPERTY(QString OutputWaveFrontFile READ getOutputWaveFrontFile WRITE setOutputWaveFrontFile)
  PYB11_PROPERTY(DataArrayPath TriangleGeometry READ getTriangleGeometry WRITE setTriangleGeometry)
```

imikejackson committed:
Correct Python binding macro

Co-Authored-By: JDuffeyBQ <jared.duffey@bluequartz.net>
